dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/EntryVersionHelper.java

```diff
@@ -237,106 +237,122 @@ default List<SourceFile> getAllSourceFiles(long workflowId, String tag, Descript
             .map(entry -> entry.getValue().getLeft()).collect(Collectors.toList());
     }
 
+    default String getDefaultVersionName() {
+        return getDAO() instanceof ToolDAO ? "latest" : "master";
+    }
+
+    /**
+     * Finds the version and entry corresponding to the specified version name and entry ID, adjusting access per the specified logged-in user.
+     * The entry will only contain the version with the specified version name, and will be evicted so no accidental db mods are possible.
+     * Throws a CustomWebApplicationException if the entry or version is not accessible.
+     * @param entryId id of the entry
+     * @param versionName name of the version
+     * @param user the logged-in user
+     * @return the version and parent entry
+     */
+    default EvictedVersionAndEntry<T> findAndCheckVersionByName(long entryId, String versionName, Optional<User> user, VersionDAO versionDAO) {
+        try {
+            String modifiedVersionName = ObjectUtils.firstNonNull(versionName, getDefaultVersionName());
+            versionDAO.enableNameFilter(modifiedVersionName);
+            T entry = findAndCheckEntryById(entryId, user);
+            Version version = entry.getWorkflowVersions().stream().filter(v -> v.getName().equals(modifiedVersionName)).findFirst().orElseThrow(
+                () -> new CustomWebApplicationException("Invalid or missing version " + modifiedVersionName + ".", HttpStatus.SC_BAD_REQUEST));
+            getDAO().evict(entry);
+            return new EvictedVersionAndEntry<>(version, entry);
+        } finally {
+            versionDAO.disableNameFilter();
+        }
+    }
+
+    /**
+     * Finds the version and entry corresponding to the specified version ID and entry ID, adjusting access per the specified logged-in user.
+     * The entry will only contain the version with the specified version ID, and will be evicted so no accidental db mods are possible.
+     * Throws a CustomWebApplicationException if the entry or version is not accessible.
+     * @param entryId id of the entry
+     * @param versionId id of the version
+     * @param user the logged-in user
+     * @return the version and parent entry
+     */
+    default EvictedVersionAndEntry<T> findAndCheckVersionById(long entryId, long versionId, Optional<User> user, VersionDAO versionDAO) {
+        try {
+            versionDAO.enableIdFilter(versionId);
+            T entry = findAndCheckEntryById(entryId, user);
+            Version version = entry.getWorkflowVersions().stream().filter(v -> v.getId() == versionId).findFirst().orElse(null);
+            if (version == null) {
+                throw new CustomWebApplicationException("Version " + versionId + " does not exist for this entry", HttpStatus.SC_BAD_REQUEST);
+            }
+            getDAO().evict(entry);
```

Comment:
Entry will not get evicted if an exception is thrown. Is that OK? I would guess so, but I don't know enough about Hibernate works to say for sure.

Hypothetical: what if the exception is thrown, but another method then searches for the entry. Will Hibernate return this cached copy? I'm guessing not, but just trying to think of an edge case.

Reply:
Good catch, I think the entry would still be in the session, I moved it to above the throw site.